docs for granular access policies (#2966)
rilldata/rill: docs/docs/develop/policies.md

---
title: "Secure: Dashboards"
description: Granular access policies for dashboards 
sidebar_label: "Secure: Dashboards"
sidebar_position: 30
---

:::caution

This is an experimental feature.

:::

Rill supports granular access policies for dashboards. You can define access policies for dashboards in the respective `dashboard_name.yaml` file under `policies` section.

> See policies configuration reference for all the possible options here.

Reviewer: Add link

## Dashboard Policies
Currently, Rill supports the following use cases for dashboards:
- Dashboard level access - Based on this condition you can restrict access to a dashboard for specific users based on user attributes.
- Row level filtering - Based on this condition you can restrict access to certain rows for a specific users based on user attributes.

> - Column level access - Based on this condition you can restrict access to certain columns in a dashboard for a specific users based on user attributes. 

Reviewer: Instead of "Column level access", can we make it "Dimension/measure level access"

Currently only dimensions defined in the dashboard definition can be used for column level access.

### User Attributes
Following user attributes are available for defining policies:
- `name` - name of the user. Example - `John Doe`    
- `email` - email of the user. Example - `john.doe@test.com`

> - `domain` - domain of the email. Example - `test.com`

Reviewer: Should use `example.com` to avoid clashes with real domains: https://en.wikipedia.org/wiki/Example.com

    ```yaml
    policy:
      has_access: "'{{ .user.admin }}' == true"
      filter: "domain = '{{ .user.domain }}'"
      exclude:
        - name: ssn
          if: "{{ .user.domain }} != 'test.com'"
    ```
   As per above ssn dimension will be excluded but all other dimensions will be included.
4. Here's one example for column inclusion.

    ```yaml
    policy:
      has_access: "true"
      filter: "domain = '{{ .user.domain }}'"
      include:
        - name: ssn
          if: "'{{ .user.admin }}' == true"
    ```
   > **_Note:_** If include is defined all other dimensions will be excluded. Only one of the
     `include`/`exclude` can be defined at a time.

> **_Note:_** `has_access` and `if` conditions on `include`/`exclude` fields needs to be valid expressions that evaluates to true/false.

## Advanced Examples
Policies are resolved in two phases - 
1. The templating engine replaces placeholder like this `{{ }}` with actual values

> 2. For `has_access` and `if` conditions, the resultant expression is evaluated to true/false, whereas for `filter` condition the resultant expression is injected in the query.

Reviewer: Clarify that it is injected as a SQL expression in the WHERE clause or the underlying query

> **_Note:_** We use sprig templating engine for templating and govaluate for expression evaluation. So any valid operator/expression they support will work here. You can refer to their documentation for more details.

> [Here](http://masterminds.github.io/sprig/) are the supported sprig functions and [here](https://github.com/Knetic/govaluate/blob/master/MANUAL.md#operators) are the supported expression evaluation operators. 

Reviewer: We actually use the Go [`text/template`](https://pkg.go.dev/text/template) templating engine with Sprig functions added to the context.

Now lets look at some advanced examples.

1. Let's say you want to restrict access to a dashboard only to admin users or users belonging to `test.com` domain.
    ```yaml 
    policy:

>       has_access: "'{{ .user.admin }}' == true || '{{ .user.domain }}' == 'test.com'"

Reviewer: 1. Are double quotes around the expression needed here?
2. Why does `{{ .user.admin }}` need to have single quotes around it? Since it's a bool, I would have thought it just resolved to `true == true`, not `'true' == true`

    ```
2. Let's say additionally we want to filter queries based on user's groups and there exist a `group` dimension in the model.
    ```yaml 
    policy:
      has_access: "'{{ .user.admin }}' == true || '{{ .user.domain }}' == 'test.com'"
      filter: "groups IN ('{{ .user.groups | join "', '" }}')"

>     ```

Reviewer: Let's remove this for now since we don't support creating user groups on admin-side yet

### Custom attributes

> Sometimes you may want to use custom attributes for defining policies other than existing user attributes. For example, you may have a `department` dimension and want to filter the dashboard view as per the users department.

Reviewer: Can we make the example `tenant_id` instead? I think that would show a more powerful use case (partner filtering)

You need to define the mapping of users to department as a separate [source](../reference/project-files/sources) in Rill. 
Then you can refer to the source mapping in the policy. 

For example, lets say we create a file named `mappings.csv` having the following data and put it in `data/` directory.
```csv
email,department
john.doe@test.com,marketing

> jane.doe@test.com,tech

Reviewer: Also use example.com

```
Now we can define a new source in `sources/mappings.yaml` file as follows:
```yaml
type: "local_file"
path: "data/mappings.csv"
``` 

Now we can refer to this source in the policy as follows:
```yaml
policy:
has_access: "'{{ .user.admin }}' == true || '{{ .user.domain }}' == 'test.com'"
filter: "department IN (SELECT department FROM mappings WHERE email = '{{ .user.email }}' )"

> ```

Reviewer: 1. Missing indentation under policy
2. Can we make `has_access` be just `true` here to focus just on the sub-query complexity?

During development, the expressions used in the policies are validated using dummy data. When a user loads a dashboard, the policies are then resolved in two phases:

1. The templating engine first replaces expressions like `{{ .user.domain }}` with actual values ([Templating reference](../reference/templating))
2. The resulting expression is then evaluated contextually:
  - The `filter` value is injected into the `WHERE` clause of the SQL queries used to render the dashboard
  - The `has_access` and `if` values are resolved to a `true` or `false` value using the expression engine ([Expressions reference](../reference/expressions))

## Examples

### Restrict dashboard access to users matching specific criteria

Let's say you want to restrict dashboard access to admin users or users whose email domain is `example.com`. Add the following clause to your dashboard's YAML:
```yaml 
policy:
  has_access: "{{ .user.admin }} == true || '{{ .user.domain }}' == 'example.com'"
```

> **_Note:_** If the `policy` section is defined and `has_access` is not, then `has_access` will default to `false`, meaning that it won't be accessible to anyone.

### Show only data from the user's own domain

> You can force the dashboard to filter all dimension and measure values by the user's domain. Assuming your dashboard's model has a `domain` column, add the following clause to your dashboard's YAML:

Reviewer: ```suggestion
You can limit the data available to the dashboard by applying a filter on the underlying data. Assuming the dashboard's underlying model has a `domain` column, adding the following clause to the dashboard's YAML will only show dimension and measure values for the current user's email domain:
```

```yaml
policy:
  has_access: true
  filter: "domain = '{{ .user.domain }}'"
```

> **_Note:_** The `filter` value needs to be valid SQL syntax for a `WHERE` clause. It will be injected into every SQL query used to render the dashboard.

### Conditionally hide a dashboard dimension or measure

You can include or exclude dimensions and measures based on a boolean expression. For example, to exclude a dimension named `ssn` for users whose email domain is not `example.com`:

```yaml
policy:
  has_access: true
  exclude:
    - name: ssn
      if: "'{{ .user.domain }}' != 'example.com'"
```

> Alternatively, you can explicitly define the dimensions and measures to include using the `include` key. It uses the same syntax as `exclude` and automatically exclude all names not explicitly defined in the list. See the [Dashboard YAML](../reference/project-files/dashboards) reference for details.

Reviewer: ```suggestion
Alternatively, you can explicitly define the dimensions and measures to include using the `include` key. It uses the same syntax as `exclude` and automatically excludes all names not explicitly defined in the list. See the [Dashboard YAML](../reference/project-files/dashboards) reference for details.
```

<!-- PENDING SUPPORT FOR USER-DEFINED USERGROUPS -->
<!--
### Filter queries based on the user's groups

Let's say additionally we want to filter queries based on user's groups and there exist a `group` dimension in the model:
```yaml 
policy:
  has_access: true
  filter: "groups IN ('{{ .user.groups | join \"', '\" }}')"
```
-->

### Advanced Example: Custom attributes

For some use cases, the built-in user attributes do not provide sufficient context to correctly restrict access. For example, a dashboard for a multi-tenant SaaS application might have a `tenant_id` column and external users should only be able to see data for the tenant they belong to.

To support this, we can ingest a separate data [source](./import-data) containing mappings of user email addresses to tenant IDs and reference it in the row-level filter.

> For example, if we have a `mappings.csv` file in the `data` directory of our Rill project:

Reviewer: ```suggestion
For example, if we have a `mappings.csv` file in the `data` directory of our Rill project with the following contents:
```
